glib-macros/src/clone_old.rs

```diff
@@ -0,0 +1,965 @@
+            Some(x) => ret.push(x),
+            None => return Err(parts.generate_error_with_span("Unexpected end 3", span)),
+        }
+    }
+    Ok(ret)
+}
+
+fn tokens_to_string(parts: impl Iterator<Item = TokenTree>) -> String {
+    let mut ret = String::new();
+    // This is used in case of "if ident" or other similar cases.
+    let mut prev_is_ident = false;
+    let handle_ident_like = |i: String, ret: &mut String, prev_is_ident: &mut bool| {
+        if *prev_is_ident {
+            ret.push(' ');
+        }
+        ret.push_str(&i);
+        *prev_is_ident = true;
+    };
+
+    for token in parts {
+        match token {
+            TokenTree::Punct(p) => {
+                prev_is_ident = false;
+                ret.push_str(&p.to_string());
+            }
+            TokenTree::Ident(i) => handle_ident_like(i.to_string(), &mut ret, &mut prev_is_ident),
+            TokenTree::Literal(l) => handle_ident_like(l.to_string(), &mut ret, &mut prev_is_ident),
+            TokenTree::Group(g) => {
+                prev_is_ident = false;
+                ret.push_str(&group_to_string(&g));
+            }
+        }
+    }
+    ret
+}
+
+fn build_closure(
+    parts: PeekableProcIter,
+    elements: Vec<ElemToClone>,
+    return_kind: Option<WrapperKind>,
+    kind: BlockKind,
+) -> TokenStream {
+    let mut body = TokenStream::new();
+
+    for el in &elements {
+        let stream: TokenStream = el
+            .to_str_after(&return_kind)
+            .parse()
+            .expect("failed to convert element after");
+        body.extend(stream.into_iter().collect::<Vec<_>>());
+    }
+    body.extend(parts.collect::<Vec<_>>());
+
+    // To prevent to lose the spans in case some errors occur in the code, we need to keep `body`!
+    //
+    // If we replaced everything that follows with a `format!`, it'd look like this:
+    //
+    // format!(
+    //     "{{\n{}\nmove |{}| {{\n{}\nlet ____ret = {{ {} }};\n____ret\n}}\n}}",
+    //     elements
+    //         .iter()
+    //         .map(|x| x.to_str_before())
+    //         .collect::<Vec<_>>()
+    //         .join("\n"),
+    //     closure,
+    //     elements
+    //         .iter()
+    //         .map(|x| x.to_str_after(&return_kind))
+    //         .collect::<Vec<_>>()
+    //         .join("\n"),
+    //     body,
+    // )
+    let mut ret: Vec<TokenTree> = vec![];
+    for el in elements {
+        let stream: TokenStream = el
+            .to_str_before()
+            .parse()
+            .expect("failed to convert element");
+        ret.extend(stream.into_iter().collect::<Vec<_>>());
+    }
+
+    // This part is creating the TokenStream using the variables that needs to be cloned (from the
+    // @weak and @strong annotations).
+    let mut inner: Vec<TokenTree> = Vec::new();
+
+    {
+        let stream = "{
+#[deprecated = \"Using old-style clone! syntax\"]
+macro_rules! clone { () => {}; }
+clone!();
+}"
```

Comment:
`clone_old.rs` and `closure_old.rs` are exactly the same as the previous versions plus the addition of this for generating deprecation warnings.

github unfortunately doesn't detect that they're basically the same files.

With the next major release (0.21) we would drop the old code.